src/pkcs11/framework-pkcs15.c

```diff
@@ -1070,9 +1070,13 @@ pkcs15_init_slot(struct sc_pkcs15_card *p15card, struct sc_pkcs11_slot *slot,
 			pin_info = NULL;
 		}
 		else   {
-			if (auth->label[0] && strncmp(auth->label, "PIN", 4) != 0)
-				snprintf(label, sizeof(label), "%.*s (%s)", (int) sizeof auth->label, auth->label, p15card->tokeninfo->label);
-			else
+			if (auth->label[0] && strncmp(auth->label, "PIN", 4) != 0) {
+				int ti_len = MIN(strlen(p15card->tokeninfo->label),
+					sizeof(label) - strlen(auth->label) - 4);
+				snprintf(label, sizeof(label), "%.*s (%.*s)",
+					(int) strlen(auth->label), auth->label,
+					ti_len, p15card->tokeninfo->label);
+			} else
```

Comment:
This still looks more complicated than it should be. What was the exact warning here?

Reply:
can you change the size of `label`?

Reply:
I have to agree with you. It would be nice if we could achieve the same somehow simpler.
We can change the size of labels, but it would
 * require dynamic allocation
 * could expand the prompt to unfriendly long line exposed to the users

The error looked this way:
```
  CC       opensc_pkcs11_la-framework-pkcs15.lo
framework-pkcs15.c: In function ‘pkcs15_init_slot’:
framework-pkcs15.c:1074:37: error: ‘%.*s’ directive output may be truncated writing up to 254 bytes into a region of size 64 [-Werror=format-truncation=]
     snprintf(label, sizeof(label), "%.*s (%s)", (int) sizeof auth->label, auth->label, p15card->tokeninfo->label);
                                     ^~~~
framework-pkcs15.c:1074:5: note: ‘snprintf’ output 4 or more bytes (assuming 258) into a destination of size 64
     snprintf(label, sizeof(label), "%.*s (%s)", (int) sizeof auth->label, auth->label, p15card->tokeninfo->label);
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

Reply:
my suggestion is to make the local buffer `label` bigger so that it can take everything you want to print into it in line 1074. Later, we still only copy 32 bytes from the local label to `token_info.label`.

Reply:
The `sc_pkcs15_tokeninfo.label` in `p15card` is dynamically allocated so I believe there is no guarantee that it will have some size. Especially no guarantee that would satisfy gcc. Though in the other cases, it should work this way.

Reply:
I think the warning is due to the fact that the receiving field of `auth->label` (`%.*s`...`(int) sizeof auth->label`) is truncated. I think there is no problem with truncating `p15card->tokeninfo->label`, because the receiving field (`%s`) doesn't specify a length.

Reply:
Yes. You are right. Extending buffer helps here too and the dynamic part as `%s` does not matter here. See the latest change.